### PR TITLE
MS17463: only show InteractiveWindow on top of interface window

### PR DESCRIPTION
**Pull request by thoys**
fixes https://highfidelity.manuscript.com/f/cases/17463/InteractiveWindows-stay-on-top-when-interface-not-focussed

## QA Test 
1. be in desktop mode
2. run script test  `scripts/developer/test/interactiveWindowTest.js`
3. a native window should pop up with the text in title bar "Native Desktop Window", and in center of the window you should see "OverlayWindowTestString"
4. switch to VR mode
5. the native window should disappear from the desktop and re-appear on the Virtual desktop with title "Virtual Desktop Window".
6. switch back to Desktop mode. Verify that you see the same result as in step 3.
7. Platform specific step:
     - On MS Windows the window should be always in front of interface and yet other windows should be able to move in front of the window. (Try moving your web browser over the other window, the interactive window should not interfere with your web browsing.)
      - On other platforms the interactive window will always overlap over any window.
7. stop the script. -> the window should disappear.
8. smoke test other already existing UI elements/windows.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12299/HighFidelity-Beta-PR13857-e08d8e9.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12299/HighFidelity-Beta-PR13857-e08d8e9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12299/HighFidelity-Beta-Interface-PR13857-e08d8e9.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12299/HighFidelity-Beta-PR13857-e08d8e9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12299/HighFidelity-Beta-Interface-PR13857-e08d8e9.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12420/HighFidelity-Beta-PR13857-efa3fa7.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12420/HighFidelity-Beta-PR13857-efa3fa7.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12420/HighFidelity-Beta-Interface-PR13857-efa3fa7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12420/HighFidelity-Beta-PR13857-efa3fa7.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12420/HighFidelity-Beta-Interface-PR13857-efa3fa7.exe

File: interface/resources/qml/InteractiveWindow.qml [hunk 1]
```diff
@@ -146,7 +146,8 @@ Windows.Window {
             Qt.WindowCloseButtonHint |
             Qt.WindowMaximizeButtonHint |
             Qt.WindowMinimizeButtonHint;
-        if ((flags & Desktop.ALWAYS_ON_TOP) === Desktop.ALWAYS_ON_TOP) {
+        // only use the always on top feature for non Windows OS
+        if (Qt.platform.os !== "windows" && (flags & Desktop.ALWAYS_ON_TOP) === Desktop.ALWAYS_ON_TOP) {
```
Comment (huffman) [inline interface/resources/qml/InteractiveWindow.qml]: The `=== Desktop.ALWAYS_ON_TOP` should be unnecessary here if flags is just a bit field.

File: libraries/ui/src/MainWindow.cpp [hunk 2]
```diff
@@ -39,6 +40,18 @@ MainWindow::~MainWindow() {
     qCDebug(uiLogging) << "Destroying main window";
 }
 
+QWindow* MainWindow::findMainWindow() {
+    auto windows = qApp->topLevelWindows();
+    QWindow* result = nullptr;
+    for (auto window : windows) {
```
Comment (huffman) [inline libraries/ui/src/MainWindow.cpp]: Can we use `const auto& window` here?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12443/HighFidelity-Beta-PR13857-c4d43b3.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12443/HighFidelity-Beta-PR13857-c4d43b3.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12443/HighFidelity-Beta-Interface-PR13857-c4d43b3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12443/HighFidelity-Beta-PR13857-c4d43b3.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12443/HighFidelity-Beta-Interface-PR13857-c4d43b3.exe

**Comment by qlilaseguinolaza** [comment]
Currently testing

**Comment by qlipdahlin** [comment]
Currently Testing

**Comment by qlipdahlin** [comment]
@conklin94122 FAIL

Entered desktop mode, added script, overlay appears properly. Entered HMD mode and overlay disappears from desktop and appears in HMD. Exited HMD and overlay appears but is behind the interface window. The interface window will go over the overlay, which based on the test plan, fails this PR.

Does not overlap other apps on Windows system.

Tested with Mac and the create menu is the only one that overlaps the designated window, the overlay will overlap other apps.

Create menu no longer overlaps other apps and primarily overlaps most interface windows. Console still overlaps other apps.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12670/HighFidelity-Beta-PR13857-198ec29.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12670/HighFidelity-Beta-PR13857-198ec29.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12670/HighFidelity-Beta-Interface-PR13857-198ec29.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12670/HighFidelity-Beta-PR13857-198ec29.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12670/HighFidelity-Beta-Interface-PR13857-198ec29.dmg

**Comment by qligmaloney** [comment]
Currently testing

**Comment by qligmaloney** [comment]
FAIL @conklin94122 Crash occurs when Running Scripts window is opened for the first time on a clean install. After the first crash, interface runs as expected. Crash happens on both Windows 10 and OSX. Native Window displays as expected after running script on both desktop and HMD platforms. Native window always displays over the top of interface. Other windows display over the top of the Native Window

[PR13857.zip](https://github.com/highfidelity/hifi/files/2354498/PR13857.zip)

**Comment by qligmaloney** [comment]
PASS @conklin94122 After further investigation, I do not believe the crashes originally witnessed are specific to this PR. Backtrace grouped the crashes together with the crashing recorded in MS18044 (linked below) meaning this has been reported prior to the release of this PR. QA Failed tag can be removed as the PR originally passed the test plan, but failed the smoke test for crashing.

**Pull request merged**